### ISSUE TITLE
thoughts about DSA key generation

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#25316]
Finally, I have enough spare time to modernize key generation in PKIX-SSH.
Remark: limited space time=long development time frame=do not touch working code

Introduction:
This is part of process started long time ago with EVP based sing/verify (otherwise FIPS 1.2 module refuse operation), EVP compatible key attribute management (opaque data structures), use PKEY instead "plain" keys (follow deprecation of "plain" key). Serialization is postponed as requires key methods to be replaced by providers.
Key generation that uses OpenSSL utilities was changed to use genpkey before to change program code. It was preferred to be activated if OpenSSL is 1.1 at least although is available since 1.0.

---
**Cases**
(FOO_generate_key vs EVP_PKEY_keygen)

**EVP_PKEY_keygen silence**
It seems to me EVP_PKEY_keygen is too silent when key generation fail. Some implementation tests shows that function (method) fail but error queue is empty. As result is not clear reason for failure.
What about failure to be described in error queue?

---
**paramgen_init() vs keygen_init()**
It seems to me EC key could be generated only with use of keygen_init() and paramgen. This is not case for DSA key. First with must create "param" pkey and then from this key to create context for generation.
Tho step process is not document well in manual page. Demos are just fine, but documentation is not so precise.
May be this is error in program code.

---
**N-bit q in DSA**
Secure shell white papers define use of DSA as public-key algorithms. Use of plain key is based on FIPS-186-2 ("Digital Signature Standard (DSS)") with sha-1. Only 1024-bit key is allowed, and standard defines q as 160-bit.
Old function DSA_generate_parameters use only key bits. Size of q is not managed. Unfortunately this is not case for keygen().

**224-bits Q**
According to documentation 224 is default for Q. This is not exact. Manual page history section lacks information that default key sizes and indirectly Q are changed in "patch" versions! Note this "patch" version.
Indirectly such changes add functional regression as programs code must set explicitly Q parameter.

**manage Q-bits**
According to documentation either ...paramgen_q_bits or ...paramgen_md could be used. It seems to me digest function dies not work at all.
Remark: This is the only case when I see error message returned for cryptographic library when generation fail: sshkey_generate: crypto message: 'error:0308010C:digital envelope routines::unsupported:Global default library context, Algorithm (xxxx : 0), Properties (<null>)' where xxxx is something binary i.e., looks like data read from random memory.
May be solution is just to document that digest function is broken.
Remark: as is not reported on time there is no need to change code - looks like never used functionality. In addition, DSA is in process of dropping form secure shell as key size is weak, CA certificate bundle lacks DSA, exist legacy enterprise Linux with broken DSA X.509 certificate support.

---
**Legacy**
May be is time to move DSA key to legacy provider.

**Comment by t8m** [comment]
> **Legacy**
> May be is time to move DSA key to legacy provider.

This cannot be done earlier than in 4.0.

**Comment by petrovr** [comment]
> As DSA is in practice an obsolete algorithm ...

Yep this is the key message i.e., candidate for legacy provider.

Issues mentioned above just shows that is rarely used.

**Comment by petrovr** [comment]
"...(would it be you ...) to take." Nope.

----

Request is opened after modernisation of key generation used for identities.
Remains modernisation of keys (DH or ECDH) used in key exchange.


PKIX-SSH 15.0 (12 Mar 2024) builds by default without DSA keys.
Further modernisation of DSA is stopped i.e., no key-manager, no serialisation based on properties(parameters).
This was last modification of DSA and should ensure compatibility with OpenSSL API 1.1.1.
Also PKIX-SSH 15.2 (11 Aug 2024) excludes DSA from manual pages.
Code will remain as some CISCO routers may use public-key algorithms based on DSA.

Key generation is not only planed modernisation.
My plans include modernisation of key methods, key store, and x509 queries(store) to provider API.
Modernisation must keep compatibility with ancient i.e. old and modern functionality must exist
and to be activated according capability of cryptographic library
 - E_NSS, engine <-> provider(key method and key store). In scope RSA and EC, DSA out.
 - E_LDAP, static engine -> provider(store based x509 queries). To move outside PKIX-SSH.
 - [private], engine -> provider(key method, key store, x509 queries). In scope RSA and EC.
 - PKIX-SSH PKCS#11, key method <-> key manager. In scope RSA and EC.

Above is TODO list with extras. It does not include maintenance and evolutions like attribute certificate and post quantum algorithms(preferable from provider).

----

As is visible my TODO list is full.
With other words I'm overloaded.
I do not have time to work for more projects.
No time even to enter into discussion i.e., I could respond with two or more weeks delay.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
